Regression tester regressed 052db7

[start automated regression testsuite report]

Dear all,

this is the automated regression testsuite reporting a new regression between
- new revision: 052db79e87c284d063186767b590e683bb8e89db  [regression testsuite results](https://cdash.dealii.org/index.php?project=deal.II&filtercount=1&showfilters=1&field1=revision&compare1=63&value1=052db7)
- old revision: f6081edf3194bc74d3884a06a2efbc27549fc42f  [regression testsuite results](https://cdash.dealii.org/index.php?project=deal.II&filtercount=1&showfilters=1&field1=revision&compare1=63&value1=f6081e)

Summary:
- :angry: new test failures introduced for configuration [GNU-13.1.1-master-autodetection-20230507](https://cdash.dealii.org/build/2858) on site tester-tos

I have identified the following pull requests as possible candidates:
- <!-- Fix compiler warning for macro usage in TpetraWrappers -->#16784 (authored-by: @masterleinad; merged-by: @bangerth)
- <!-- Remove/deprecate Tensor::begin/end_raw(). -->#16790 (authored-by: @bangerth; merged-by: @drwells)
- <!-- Provide a better error message. -->#16791 (authored-by: @bangerth; merged-by: @drwells)
- <!-- FEEvaluation: Ensure assertions in get_value/gradient are working -->#16789 (authored-by: @kronbichler; merged-by: @drwells)


Notes:
 - The regression reported above is a subset of the following full set of regressions compared to the baseline:
   - :angry: test failures for configuration [GNU-13.1.1-master-autodetection-20230507](https://cdash.dealii.org/build/2858) on site tester-tos
 - I will close and unpin this issue automatically once a full run is complete and compares cleanly to the baseline.
 - If closed I will reopen the issue if the testsuite run identified an additional regression compared to what I have reported so far.

[end automated regression testsuite report]

This was caused by #16790:
```
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc: In instantiation of ‘void test_tensor() [with int rank = 0; int dim = 2]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:151:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:22: error: ‘using Tensor_t = class dealii::Tensor<0, 2, double>’ {aka ‘class dealii::Tensor<0, 2, double>’} has no member named ‘begin_raw’
  106 |   for (auto it = t_a.begin_raw(); it != t_a.end_raw(); ++it)
      |                  ~~~~^~~~~~~~~
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:45: error: ‘using Tensor_t = class dealii::Tensor<0, 2, double>’ {aka ‘class dealii::Tensor<0, 2, double>’} has no member named ‘end_raw’
  106 |   for (auto it = t_a.begin_raw(); it != t_a.end_raw(); ++it)
      |                                         ~~~~^~~~~~~
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:108:22: error: ‘using Tensor_t = class dealii::Tensor<0, 2, double>’ {aka ‘class dealii::Tensor<0, 2, double>’} has no member named ‘begin_raw’
  108 |   for (auto it = t_b.begin_raw(); it != t_b.end_raw(); ++it)
      |                  ~~~~^~~~~~~~~
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:108:45: error: ‘using Tensor_t = class dealii::Tensor<0, 2, double>’ {aka ‘class dealii::Tensor<0, 2, double>’} has no member named ‘end_raw’
  108 |   for (auto it = t_b.begin_raw(); it != t_b.end_raw(); ++it)
      |                                         ~~~~^~~~~~~
In file included from /srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/symmetric_tensor.h:25,
                 from /srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/differentiation/sd/symengine_tensor_operations.h:22,
                 from /srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/differentiation/sd/symengine_optimizer.h:42,
                 from /srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/differentiation/sd.h:25,
                 from /srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:18:
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘Number* dealii::Tensor<rank_, dim, Number>::begin_raw() [with int rank_ = 2; int dim = 2; Number = double]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:31:   required from ‘void test_tensor() [with int rank = 2; int dim = 2]’
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:153:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1507:23: error: static assertion failed: This function is only available for rank-1 tensors because higher-rank tensors may not store their elements in a contiguous array.
 1507 |   static_assert(rank_ == 1,
      |                 ~~~~~~^~~~
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1507:23: note: the comparison reduces to ‘(2 == 1)’
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘Number* dealii::Tensor<rank_, dim, Number>::end_raw() [with int rank_ = 2; int dim = 2; Number = double]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:52:   required from ‘void test_tensor() [with int rank = 2; int dim = 2]’
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:153:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1537:23: error: static assertion failed: This function is only available for rank-1 tensors because higher-rank tensors may not store their elements in a contiguous array.
 1537 |   static_assert(rank_ == 1,
      |                 ~~~~~~^~~~
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1537:23: note: the comparison reduces to ‘(2 == 1)’
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘Number* dealii::Tensor<rank_, dim, Number>::begin_raw() [with int rank_ = 3; int dim = 2; Number = double]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:31:   required from ‘void test_tensor() [with int rank = 3; int dim = 2]’
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:154:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1507:23: error: static assertion failed: This function is only available for rank-1 tensors because higher-rank tensors may not store their elements in a contiguous array.
 1507 |   static_assert(rank_ == 1,
      |                 ~~~~~~^~~~
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1507:23: note: the comparison reduces to ‘(3 == 1)’
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘Number* dealii::Tensor<rank_, dim, Number>::end_raw() [with int rank_ = 3; int dim = 2; Number = double]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:52:   required from ‘void test_tensor() [with int rank = 3; int dim = 2]’
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:154:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1537:23: error: static assertion failed: This function is only available for rank-1 tensors because higher-rank tensors may not store their elements in a contiguous array.
 1537 |   static_assert(rank_ == 1,
      |                 ~~~~~~^~~~
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1537:23: note: the comparison reduces to ‘(3 == 1)’
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘Number* dealii::Tensor<rank_, dim, Number>::begin_raw() [with int rank_ = 4; int dim = 2; Number = double]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:31:   required from ‘void test_tensor() [with int rank = 4; int dim = 2]’
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:155:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1507:23: error: static assertion failed: This function is only available for rank-1 tensors because higher-rank tensors may not store their elements in a contiguous array.
 1507 |   static_assert(rank_ == 1,
      |                 ~~~~~~^~~~
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1507:23: note: the comparison reduces to ‘(4 == 1)’
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘Number* dealii::Tensor<rank_, dim, Number>::end_raw() [with int rank_ = 4; int dim = 2; Number = double]’:
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:106:52:   required from ‘void test_tensor() [with int rank = 4; int dim = 2]’
/srv/temp/testsuite-icEUnkf1/dealii/tests/symengine/symengine_tensor_operations_04.cc:155:22:   required from here
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1537:23: error: static assertion failed: This function is only available for rank-1 tensors because higher-rank tensors may not store their elements in a contiguous array.
 1537 |   static_assert(rank_ == 1,
      |                 ~~~~~~^~~~
/srv/temp/testsuite-icEUnkf1/dealii/include/deal.II/base/tensor.h:1537:23: note: the comparison reduces to ‘(4 == 1)’
gmake[3]: *** [CMakeFiles/symengine.symengine_tensor_operations_04.debug.dir/build.make:76: CMakeFiles/symengine.symengine_tensor_operations_04.debug.dir/symengine_tensor_operations_04.cc.o] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:9387: CMakeFiles/symengine.symengine_tensor_operations_04.debug.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:9420: CMakeFiles/symengine.symengine_tensor_operations_04.debug.test.dir/rule] Error 2
gmake: *** [Makefile:4541: symengine.symengine_tensor_operations_04.debug.test] Error 2
Test symengine/symengine_tensor_operations_04.debug: BUILD
```

I will try to fix these tests later today, but am short of time. If someone wants to beat me to it, that's ok too.

--- a/tests/symengine/symengine_tensor_operations_04.cc
+++ b/tests/symengine/symengine_tensor_operations_04.cc
@@ -102,11 +102,20 @@ test_tensor()
   using Tensor_SD_number_t = Tensor<rank, dim, SD_number_t>;
   using Tensor_t           = Tensor<rank, dim, double>;
 
+  // Fill two tensors with ones and twos, respectively.
   Tensor_t t_a, t_b;
-  for (auto it = t_a.begin_raw(); it != t_a.end_raw(); ++it)
-    *it = 1.0;
-  for (auto it = t_b.begin_raw(); it != t_b.end_raw(); ++it)
-    *it = 2.0;
+  if constexpr (rank > 0)
+    {
+      for (unsigned int i = 0; i < Tensor_t::n_independent_components; ++i)
+        t_a[Tensor_t::unrolled_to_component_indices(i)] = 1.0;
+      for (unsigned int i = 0; i < Tensor_t::n_independent_components; ++i)
+        t_b[Tensor_t::unrolled_to_component_indices(i)] = 2.0;
+    }
+  else
+    {
+      t_a = 1.0;
+      t_b = 2.0;
+    }
 
   const Tensor_SD_number_t symb_t_a =
     SD::make_tensor_of_symbols<rank, dim>("a");

